beetsplug/export.py

```diff
@@ -0,0 +1,145 @@
+            'default_format': 'json',
+            'json': {
+                # json module formatting options
+                'formatting': {
+                    'ensure_ascii': False,
+                    'indent': 4,
+                    'separators': (',', ': '),
+                }
+            },
+            # TODO: Use something like the edit plugin
+            # 'item_fields': []
+        })
+
+    def commands(self):
+        # TODO: Add option to use albums
+
+        cmd = ui.Subcommand('export', help=u'export data from beets')
+        cmd.func = self.run
+        cmd.parser.add_option(
+            u'-l', u'--library', action='store_true',
+            help=u'show library fields instead of tags',
+        )
+        cmd.parser.add_option(
+            u'--append', action='store_true', default=False,
+            help=u'if should append data to the file',
+        )
+        cmd.parser.add_option(
+            u'-i', u'--include-keys', default=[],
+            action='append', dest='included_keys',
+            help=u'comma separated list of keys to show',
+        )
+        cmd.parser.add_option(
+            u'-p', u'--path',
+            help=u'path for the output file. If not given, will print the data'
+        )
+        return [cmd]
+
+    def run(self, lib, opts, args):
+
+        file_path = opts.path
+        file_format = self.config['default_format'].get(str)
+        file_mode = 'a' if opts.append else 'w'
+        format_options = self.config[file_format]['formatting'].get(dict)
+
+        export_format = ExportFormat.factory(
+            file_format, **{
+                'file_path': file_path,
+                'file_mode': file_mode
+            }
+        )
+
+        items = []
+        data_collector = library_data if opts.library else tag_data
+
+        included_keys = []
+        for keys in opts.included_keys:
+            included_keys.extend(keys.split(','))
+        key_filter = make_key_filter(included_keys)
+
+        for data_emitter in data_collector(lib, ui.decargs(args)):
+            try:
+                data, item = data_emitter()
+            except (mediafile.UnreadableFileError, IOError) as ex:
+                self._log.error(u'cannot read file: {0}', ex)
+                continue
+
+            data = key_filter(data)
+            items += [data]
+
+        export_format.export(items, **format_options)
+
+
+class ExportFormat(object):
+    """The output format type"""
+
+    @classmethod
+    def factory(self, type, **kwargs):
+        if type == "json":
+            if kwargs['file_path']:
+                return JsonFileFormat(**kwargs)
+            else:
+                return JsonPrintFormat()
+        raise NotImplemented("")
+
+    def export(self, data, **kwargs):
+        raise NotImplemented("")
+
+
+class JsonPrintFormat(ExportFormat):
+    """Outputs to the console"""
+
+    def export(self, data, **kwargs):
+        ui.print_(json.dumps(data, **kwargs))
```

Comment:
It's possible this could be shorter if we just used the same `json.dump` call as below, just with `sys.stdout` instead of an opened file.